### PR TITLE
add metadata information to PROTOCOL.md

### PR DESCRIPTION
**Pull request by k0tk267**
## Description
I add metadata information to PROTOCOL.md because the link TODO is not available.

This PR is link to [#1250](https://github.com/delta-io/delta/issues/1250)

## How was this patch tested?
Did't do test. This PR is document only changes.

## Does this PR introduce _any_ user-facing changes?
No

### CODE REVIEW

File: PROTOCOL.md [hunk 1]
```diff
@@ -881,7 +881,16 @@ Field Name | Description
 name| Name of this (possibly nested) column
 type| String containing the name of a primitive type, a struct definition, an array definition or a map definition
 nullable| Boolean denoting whether this field can be null
-metadata| A JSON map containing information about this column. Keys prefixed with `Delta` are reserved for the implementation. See [TODO](#) for more information on column level metadata that clients must handle when writing to a table.
+metadata| A JSON map containing information about this column. Keys prefixed with `Delta` are reserved for the implementation. See [metadata](#metadata) for more information on column level metadata that clients must handle when writing to a table.
+
+### metadata
+This field include column mapping information to avoid any column naming restrictions.
+
+Field Name | Description
```
Comment (scottsand-db) [inline PROTOCOL.md:889]: You can give an example of what types of keys/values are stored in the metadata. which is just a map of String->JSON-encoded value.

For example, the Column Mapping section gives some examples. he physical name is stored as part of the column metadata with the key delta.columnMapping.physicalName. The column id is stored within the metadata with the key delta.columnMapping.id. The column mapping is governed by the table property delta.columnMapping.mode and can be one of none, id, and name.
```

Generated columns also gives an example. The metadata for a column in the table schema MAY contain the key delta.generationExpression.
```
The metadata for a column in the table schema MAY contain the key delta.generationExpression.
```
The physical name is stored as part of the column metadata with the key delta.columnMapping.physicalName. The column id is stored within the metadata with the key delta.columnMapping.id. The column mapping is governed by the table property delta.columnMapping.mode and can be one of none, id, and name.
```

**Review by scottsand-db — CHANGES_REQUESTED** [review]
Welcome to delta! Thanks for contributing!

Left some feedback. Thanks for helping us improve our docs and protocol!

Let me know if you have any questions.

**Comment by k0tk267** [comment]
I fix document, but don't have enough confidence to cover all metadata types.
So, please teach me when I miss something.
In addition, I tried to add examples like bellow.
```
{
  "type" : "struct",
  "fields" : [ {
    "name" : "asset",
    "type" : "string",
    "nullable" : true,
    "metadata" : {
      "delta.columnMapping.id": 1,
      "delta.columnMapping.physicalName": "col-b96921f0-2329-4cb3-8d79-184b2bdab23b"
      "delta.identity.allowExplicitInsert": true,
      "delta.invariants": "{\"expression\": { \"expression\": \"x > 3\"} }",
      "delta.generationExpression": "???"
      ""
    }
  }
}
```
However, I did't know the example of `delta.generationExpression`.
So, if I need to add these examples, please teach me the example of `delta.generationExpression` or the reference of this.

File: PROTOCOL.md [hunk 2]
```diff
@@ -903,6 +903,17 @@ type| Always the string "map".
...
 
+### Column Metadata
+A column metadata stores various information about the column.
+For example, this MAY contain some keys like [`delta.columnMapping`](#column-mapping) or [`delta.generationExpression`](#generated-columns).  
+Field Name | Description
+-|-
+delta.columnMapping| This is to avoid any column naming restrictions, and to support the renaming and dropping of columns without having to rewrite all the data. The more information is [Column Mapping](#column-mapping)
+delta.identity.xxx| This is for defining identity columns. The more information is [Identity Columns](#identity-columns)
+delta.invariants| Always JSON string. The more information is [Column Invariants](#column-invariants)
+delta.generationExpression| Alaways parsed as SQL expression. The more information is [Generated Columns](#generated-columns)
```
Comment (tdas) [inline PROTOCOL.md:914]:
```suggestion
delta.generationExpression| Always parsed as SQL expression. The more information is [Generated Columns](#generated-columns)
```

File: PROTOCOL.md [hunk 3]
```diff
@@ -903,6 +903,17 @@ type| Always the string "map".
 keyType| The type of element used for the key of this map, represented as a string containing the name of a primitive type, a struct definition, an array definition or a map definition
 valueType| The type of element used for the key of this map, represented as a string containing the name of a primitive type, a struct definition, an array definition or a map definition
 
+### Column Metadata
+A column metadata stores various information about the column.
+For example, this MAY contain some keys like [`delta.columnMapping`](#column-mapping) or [`delta.generationExpression`](#generated-columns).  
+Field Name | Description
+-|-
+delta.columnMapping| This is to avoid any column naming restrictions, and to support the renaming and dropping of columns without having to rewrite all the data. The more information is [Column Mapping](#column-mapping)
```
Comment (tdas) [inline PROTOCOL.md:911]:
```suggestion
delta.columnMapping.*| These keys are used to store information about the mapping between the logical column name to  the physical name. See [Column Mapping](#column-mapping) for details.
```

there are multiple metadata entries with that columnMapping prefix

**Comment by k0tk267** [comment]
@tdas 
Thank you for your review !
I checked and fixed it !

File: PROTOCOL.md [hunk 4]
```diff
@@ -903,6 +903,17 @@ type| Always the string "map".
 keyType| The type of element used for the key of this map, represented as a string containing the name of a primitive type, a struct definition, an array definition or a map definition
 valueType| The type of element used for the key of this map, represented as a string containing the name of a primitive type, a struct definition, an array definition or a map definition
 
+### Column Metadata
+A column metadata stores various information about the column.
+For example, this MAY contain some keys like [`delta.columnMapping`](#column-mapping) or [`delta.generationExpression`](#generated-columns).  
+Field Name | Description
+-|-
+delta.columnMapping.*| These keys are used to store information about the mapping between the logical column name to  the physical name. See [Column Mapping](#column-mapping) for details.
+delta.identity.xxx| This is for defining identity columns. The more information is [Identity Columns](#identity-columns)
```
Comment (tdas) [inline PROTOCOL.md:912]: super minor: make this consistent with the format of column mapping
```suggestion
delta.identity.*| These keys are for defining identity columns. See [Identity Columns](#identity-columns) for details.
```

File: PROTOCOL.md [hunk 5]
```diff
@@ -903,6 +903,17 @@ type| Always the string "map".
...
+delta.invariants| Always JSON string. The more information is [Column Invariants](#column-invariants)
+delta.generationExpression| Always parsed as SQL expression. The more information is [Generated Columns](#generated-columns)
```
Comment (tdas) [inline PROTOCOL.md:914]:
```suggestion
delta.generationExpression| SQL expression string. See [Generated Columns](#generated-columns) for details.
```

**Review by tdas — CHANGES_REQUESTED** [review]
Overall thank you for improve the quality of the Delta protocol doc :)

**Comment by k0tk267** [comment]
I have only committed to your review ... 
I'm sorry to bother you.

**Review by tdas — APPROVED** [review]
LGTM. will merge.

**Pull request closed without merging**